keel-core/src/main/kotlin/com/netflix/spinnaker/keel/constraints/TimeWindowConstraintEvaluator.kt

```diff
@@ -0,0 +1,203 @@
+package com.netflix.spinnaker.keel.constraints
+
+import com.netflix.spinnaker.keel.api.DeliveryArtifact
+import com.netflix.spinnaker.keel.api.DeliveryConfig
+import com.netflix.spinnaker.keel.api.TimeWindowConstraint
+import com.netflix.spinnaker.keel.constraints.ConstraintEvaluator.Companion.getConstraintForEnvironment
+import com.netflix.spinnaker.keel.exceptions.InvalidConstraintException
+import com.netflix.spinnaker.kork.dynamicconfig.DynamicConfigService
+import org.springframework.stereotype.Component
+import java.text.ParsePosition
+import java.time.Clock
+import java.time.DayOfWeek
+import java.time.format.DateTimeFormatter
+import java.time.format.TextStyle
+import java.util.Locale
+import java.util.TimeZone
+import java.util.TimeZone.getTimeZone
+
+/**
+ * An environment promotion constraint to gate promotions to time windows.
+ * Multiple windows can be specified.
+ *
+ * Hours and days parameters support a mix of ranges and comma separated values.
+ *
+ * Days can be specified using full or short names according to the default JVM locale.
+ *
+ * A timezone can be set in the constraint and a system-wide default can by set via
+ * the dynamic property "default.time-zone" which defaults to UTC.
+ *
+ * Example env constraint:
+ *
+ * constraints:
+ *  - type: time-window
+ *    windows:
+ *      - days: Monday,Tuesday-Friday
+ *        hours: 11-16
+ *        tz: America/Los_Angeles
+ *      - days: Wed
+ *        hours: 12-14
```

Comment:
Any opinions on 24hr time vs `12pm-2pm`? Latter seems a little more readable, but not sure whether it's "better" from an overall perspective.

Reply:
next you'll want us to use Fahrenheit

Reply:
Kelvin, obviously.

Reply:
Am and pm, how very quaint.

Reply:
I'm delighted with 24hour

Reply:
Alright alright, if everybody prefers 24h then let’s go with that. I’m coming from the perspective that for really _any_ UI we and other engineers use day to day, times get mapped to 12h because that’s typically easier to visually parse at a glance. I generally see these YAML configs as more similar in their UX needs to UIs vs. actual programming code.